From: Petr Mladek <pmladek@suse.cz>
Date: Wed, 30 Jul 2014 11:34:15 +0200
Subject: kgr: remove patch from global list when being removed
Patch-mainline: submitted for review
References: fate#313296

There is a problem when skipped patches are applied to modules.
We must ignore patches that are being removed. But this does not work now
because the patch is in the global list until it is fully removed.

This commit solves the problem by removing the patch from the global list
immediately when the revert begins.

It means that the global list contains only finalized patches.

Signed-off-by: Petr Mladek <pmladek@suse.cz>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 kernel/kgraft.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/kernel/kgraft.c b/kernel/kgraft.c
index 2396c6484ba7..992f8ebb7ced 100644
--- a/kernel/kgraft.c
+++ b/kernel/kgraft.c
@@ -158,7 +158,6 @@ static void kgr_finalize(void)
 	free_percpu(kgr_patch->irq_use_new);
 
 	if (kgr_revert) {
-		list_del(&kgr_patch->list);
 		kgr_refs_dec();
 		module_put(kgr_patch->owner);
 	} else
@@ -504,7 +503,9 @@ int kgr_modify_kernel(struct kgr_patch *patch, bool revert)
 	kgr_in_progress = true;
 	kgr_patch = patch;
 	kgr_revert = revert;
-	if (!revert)
+	if (revert)
+		list_del(&kgr_patch->list);
+	else
 		kgr_refs_inc();
 	mutex_unlock(&kgr_in_progress_lock);
 
-- 
2.1.0

